creator-node/src/services/sync/primarySyncFromSecondary.js

```diff
@@ -487,5 +561,175 @@ async function filterOutAlreadyPresentDBEntries({
+    data: { numFilteredEntries: filteredEntries.length }
+  })
+
   return filteredEntries
 }
+
+/**
+ * Given fetchedEntries, filters out entries already present in local DB
+ *
+ * @notice This function could potentially take a long time as it fetches every single row in `tableInstance` for `cnodeUserUUID`
+ *    Memory consumption is minimized by batching this call, but it can still take a long time for users with lots of data
+ *
+ * @param {Object} param
+ * @param {string} param.cnodeUserUUID
+ * @param {*} param.tableInstance Sequelize model instance to query
+ * @param {Object[]} param.fetchedEntries array of entry objects to filter out
+ * @param {*} param.transaction Sequelize transaction
+ * @param {string[]} comparisonFields fields to use for equality comparison
+ * @returns {Object[]} filteredEntries filtered version of fetchedEntries
+ */
+async function filterOutAlreadyPresentDBEntries({
+  cnodeUserUUID,
+  tableInstance,
+  fetchedEntries,
+  transaction,
+  comparisonFields,
+  decisionTree
+}) {
+  decisionTree.recordStage({
+    name: 'filterOutAlreadyPresentDBEntries() Begin',
+    data: {
+      tableName: tableInstance.name,
+      numFetchedEntries: fetchedEntries.length
+    },
+    log: true
+  })
+
+  // Timestamp is attached to prevent race conditions where same user is processed twice
+  const timestamp = Date.now()
+  const LOCAL_DB_ENTRIES_SET_KEY = `FILTER_OUT_ALREADY_PRESENT_DB_ENTRIES_LOCAL_ENTRIES_SET:::${cnodeUserUUID}:::${timestamp}`
+  const FETCHED_ENTRIES_SET_KEY = `FILTER_OUT_ALREADY_PRESENT_DB_ENTRIES_FETCHED_ENTRIES_SET:::${cnodeUserUUID}:::${timestamp}`
+  const UNIQUE_FETCHED_ENTRIES_SET_KEY = `FILTER_OUT_ALREADY_PRESENT_DB_ENTRIES_UNIQUE_FETCHED_ENTRIES_SET:::${cnodeUserUUID}:::${timestamp}`
+  await redis.del(
+    LOCAL_DB_ENTRIES_SET_KEY,
+    FETCHED_ENTRIES_SET_KEY,
+    UNIQUE_FETCHED_ENTRIES_SET_KEY
+  )
+  decisionTree.recordStage({
+    name: 'filterOutAlreadyPresentDBEntries() Ensure clean starting redis state',
+    log: true
+  })
+
+  /** Store all fetched entries into redis set (have to serialize arrays) */
+  const fetchedEntriesComparable = fetchedEntries.map((entry) =>
+    JSON.stringify(_.pick(entry, comparisonFields))
+  )
+  const numFetchedEntriesAdded = await redis.sadd(
+    FETCHED_ENTRIES_SET_KEY,
+    fetchedEntriesComparable
+  )
+  if (numFetchedEntriesAdded !== fetchedEntries.length) {
+    throw new Error(
+      `Failed to add all entries to redis set for ${FETCHED_ENTRIES_SET_KEY}`
+    )
+  }
+  decisionTree.recordStage({
+    name: 'filterOutAlreadyPresentDBEntries() Set FETCHED_ENTRIES_SET_KEY',
+    log: true
+  })
+
+  /** Store all local DB entries into redis set */
+  const limit = DB_QUERY_LIMIT
+  let offset = 0
+  let complete = false
+  let numLocalEntriesAdded = 0
+  while (!complete) {
+    let localEntries = await tableInstance.findAll({
+      where: { cnodeUserUUID },
+      limit,
+      offset,
```

Comment:
I like abstracting the table here. only concern is using limit+offset as the table grows, especially since syncs already take a while. do you think it's worth changing to cursor based pagination?

Reply:
ya i'll try and change you right

Reply:
done